### ISSUE TITLE
Project proposal: GPflowOpt

### ISSUE THREAD

**Issue by javdrher** [GPflow/GPflow#397]
Following the discussion in #324 , here a project proposal for GPflowOpt. While I was listing the responses to the questions raised in order to describe the project, it got a bit long so I made a document (attached) describing the goals, and comparison to existing alternatives.
[gpflowopt.pdf.zip](https://github.com/GPflow/GPflow/files/914704/gpflowopt.pdf.zip)
Internally, we had some code available for BO with GPflow, I took parts of it and put up a small demo with a notebook containing a constrained EI example (https://github.com/javdrher/optdemo). We would start based on this, test it fully and initiate proper documentation, rather than the brief comments now. Only then expand further to increase the number of supported features such as more complex domains (and optimizers), better designs, more acquisition functions etc.

Of course, we are available to discuss this on Github or over Skype. Additionally I can head over to the UK for a face-to-face discussion, should that be helpful.

**Comment by jameshensman** [comment]
Well, colour me impressed, this looks very good and well thought out. 

That notebook is great. 

One important edit: Ipython notebooks for demoing the functionality should be a pretty core part of the work, not under "Misc (optional)". I think you should make documentation from the notebooks in the same way as GPflow does. There's a small set up cost, but it adds huge value to the project. 

It looks like one edit you might like to make to GPflow would be the ability to have a model with 'no data' in it. I think that's a reasonable reqyuirement and something we should consider doing. Most of the functionality could be done in the GPmodel Base class, then it's as one-liner in child classes to deal with the case of `self.X is None`. 

Another convenience function that you might want in GPflow is predictions averaged over an MCMC posterior. We don't include that at the moment, perhaps we should, it would be useful here and in other cases too. 

I think there are some better space filling designs that you could use for initial points, including Sobol sequences and draws from a DPP. Not to worry: the fact that you have incorporated this kind of option into the design is great. 

I'm intrigued to see how the Domain class works out. I think that having Discrete domains is going to be really important. A pertinent question how you intent to optimize the acquisition fn over the discrete domains: random search coupled with L-BFGS-B on the continuous domains?

**Comment by javdrher** [comment]
Documentation notebooks are definitely not optional, I agree. 

Having no data in a GP would indeed be a nice feature; the way I coded it now is that any data in the GP is not discarded, and complemented with data from a design if specified in BayesianOptimizer. That behavior would remain valid, but there would be no requirement to take care of any evaluation outside of BayesianOptimizer

Regarding MCMC; in most Bayesian Optimization papers I saw, the acquisition function is averaged over hyperparameter draws rather than the averaged first and second moments of the predictive distribution. I was thinking of adding a `hyper_draws` option to BayesianOptimizer and automatically take care of that in case its set to a value larger than one.

With regards to space filling designs, we can definitely add more. Wondering why Sobol sequences specifically? Its often raised they lose their properties if dimensionality increases. Its why I have largely overlooked them until now. I'm happy to be convinced otherwise. Actually theoretically the best designs are generated according to the MaxPro criterion (https://academic.oup.com/biomet/article-abstract/102/2/371/246859/Maximum-projection-designs-for-computer) as they extend LHD with optimal space-fillingness in all possible subspace projections. However generating them makes optimizing latin hypercubes look easy.

The domain class is another thing to improve I guess. I'm not entirely certain if deriving from Parentable is actually appropriate (the naming doesn't seem applicable). The discrete optimization is another thing to investigate. Back when we used MATLAB the GA was sometimes a reasonable option to avoid randomness or exhaustive searches. I'd definitely have to look into the availability of algorithms that can handle the discrete part. It gets even more messy for categorical parameters. I'm only aware of the approach to transform those into a set of binary indicator variables, however this implies constraining the optimization to assure only one of those variables is selected at the time. 

From earlier implementations I'm aware that the complexity can quickly explode. Bayesian Optimization but there's a wild number of potential features introducing many corner cases. I guess taking small steps will be the best option.

**Comment by javdrher** [comment]
I thought a bit about the general shape of a development roadmap:
- v0.1: support common single-objective acquisition functions, MCMC, test coverage, docs, and a functional repository (with CI). The step from that demo wouldn't be that big (mostly test, docs and setup). 
- v0.2: support multi-objective setting, and PES.
- v0.3: Broaden the support for more complex optimization domains (discrete/categorical) and their optimization. This increases the range of problems the library would be applicable to.
- v0.4: Proceed into the "meta-learning" (such as freeze-thaw), and the more optional features. 
Along the development we'd hope to attract feedback and input which might of course steer some choices, especially for this phase. 

Intermediate releases would mostly cover supporting new GPflow versions, or incremental addition of new features.
 
Additional general thought: Tensorboard could be a very easy way to track the optimization process (as in #229). I'd have to look a bit into how it works to have a better idea what would be needed.

**Comment by Bonnevie** [comment]
@jameshensman: regarding the empty GPs, I have long thought that it would be nice to have a convenient way to draw random samples from a GP prior.

**Comment by jameshensman** [comment]
@javdrher , you now have a new blank repo at GPflow/GPflowOpt. I think you should have full acesss that repo, though you may have to let me know if you need other members to be added to the GPflow org.

A couple of notes about this new repo follow.

 - I've added a 'pre-release' note at the top of the repo: please don't modify this until GPflowOpt meets the already-discussed standards in terms of testing and documentation. Once you think the project is ready for other people to use, we can remove or modify that notice together. 

 - I think there is a small worry that GPflowOpt becomes 'team Ghent': please be open to collaboration in the same spirit as GPflow.

Thanks for taking on leadership of this project: I'm looking forward to using GPflowOpt in my own work!

Best of luck,

James Hensman (also on behalf of @alexggmatthews )

**Comment by icouckuy** [comment]
Thanks @jameshensman ! Of course we would like very much for people to contribute. The goal is to discuss things in the open (github issues) and have the project stable at all time to encourage this.

**Comment by 1Reinier** [comment]
Interested in this too
